src/win/tty.c

```diff
@@ -2255,25 +2250,67 @@ int uv_tty_reset_mode(void) {
   return 0;
 }
 
+static int uv__set_vtp(HANDLE handle, BOOL vtp) {
+  DWORD dwMode = 0;
+
+  if (!GetConsoleMode(handle, &dwMode)) {
+    return uv_translate_sys_error(GetLastError());
+  }
+
+  if (vtp) {
+    dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+  } else {
+    dwMode &= ~(ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+  }
+  if (!SetConsoleMode(handle, dwMode)) {
+    return uv_translate_sys_error(GetLastError());
+  }
+
+  return 0;
+}
+
 /* Determine whether or not this version of windows supports
  * proper ANSI color codes. Should be supported as of windows
  * 10 version 1511, build number 10.0.10586.
  */
 static void uv__determine_vterm_state(HANDLE handle) {
-  DWORD dwMode = 0;
-
-  if (!GetConsoleMode(handle, &dwMode)) {
-    uv__vterm_state = UV_UNSUPPORTED;
+  if (uv__set_vtp(handle, TRUE)) {
+    uv__vterm_state = UV_TTY_LEGACY;
     return;
   }
 
-  dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
-  if (!SetConsoleMode(handle, dwMode)) {
-    uv__vterm_state = UV_UNSUPPORTED;
-    return;
+  uv__vterm_state = UV_TTY_VTP;
+}
+
+int uv_tty_set_vterm_state(uv_tty_t* tty, uv_tty_vtermstate_t state) {
+  int ret = 0;
+
+  if (!(tty->flags & UV_HANDLE_WRITABLE)) {
+    return UV_EINVAL;
   }
 
-  uv__vterm_state = UV_SUPPORTED;
+  uv_sem_wait(&uv_tty_output_lock);
+  switch (state) {
+    case UV_TTY_AUTODETECT:
+      uv__determine_vterm_state(tty->handle);
```

Comment:
I think it might be better to defer the console mode change until the TTY gets next used, and to accept the user command even if we can't verify it. Thus maybe replace the switch with:
```suggestion
  if (state != UV_TTY_AUTODETECT || uv__vterm_state == UV_TTY_VTP)
    ret = uv__set_vtp(tty->handle, state == UV_TTY_VTP);
  uv__vterm_state = state;
```

Reply:
Since it is possible that `uv_tty_set_vterm_state(tty, UV_TTY_AUTODETECT)` may be called after calling `uv_tty_init`, then we need to check `UV_TTY_AUTODETECT` at the next location in `uv_tty_write_bufs`. I think that automatic detection should be done with `uv_tty_set_vterm_state` rather than that.

https://github.com/libuv/libuv/blob/ee48e6e4dce77253421514a35b0fa3efd831bf25/src/win/tty.c#L1660